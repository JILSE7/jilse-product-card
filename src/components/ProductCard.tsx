import styles from '../styles/styles.module.css';


import React, { createContext } from 'react';
import useCount from '../hooks/useCount';
import { IProductContext, IPropsProductCardComponent } from '../interfaces/ProductInterfaces';



export const ProductContext = createContext({} as IProductContext);
const {Provider} = ProductContext;

export const ProductCard = ({product, children, className, style, onChange, value, initialValues}:IPropsProductCardComponent) => {
    
  

  const {counter, increasyBy, maxCount, isMaxCountReached, reset} = useCount({onChange, product,value, initialValues});

  return (
    <Provider value={{counter, increasyBy, product, maxCount}}>
      
        <div className={`${styles.productCard} ${className}`} style={style}>

            {
           
                  children({count:counter,increasyBy,isMaxCountReached,maxCount: maxCount!, reset, product})
               
            }

        </div>

    </Provider>
  )
}

//*OTRA FORMA DE EXPORTAR LOS COMPONENTES PARA QUE SE MIREN DE LA SIGUIENTE FORMA
//?PASAMOS DE ESTO, QUE NO ESTA MAL, SOLO ES UNA FORMA 
{/* <ProductCard product={product}>
  <ProductImage/>
  <ProductTitle title="Coffe Cup"/>
  <ProducButtons counter={counter} increasyBy={increasyBy}/>
</ProductCard> */}

//?PASAMOS A PONERLE MAS PROPIEDADES AL COMPONENTE
 //* <ProductCard product={product}>
      //?<ProductCard.Image/>
      //?<ProductCard.Title title="Coffe Cup"/>
      //?<ProductCard.Buttons counter={counter} increasyBy={increasyBy}/>
//* </ProductCard> 

//*ASIGNAMOS LAS PROPIEDADES


