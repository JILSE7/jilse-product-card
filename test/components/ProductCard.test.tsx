

import React from 'react';
//import * as ReactDOM from 'react-dom';
import rendered from 'react-test-renderer';
import  ProductCard  from '../../src/components';

import { product2 } from '../data/products';


const {act} = rendered;

describe('Pruebas en el product card', () => {

    test('debe mostrar el componente correctamente', () => { 
        
        const wrapper = rendered.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <h1>Product Card</h1>
                    )
                }
            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();

    })

    test("Debe incrementar el contador", () => {

        const wrapper = rendered.create(
            <ProductCard product={product2}>
                {
                    ({count, increasyBy}) => (
                        <>
                            <h1>Product Card</h1>
                            <span>{count}</span>
                            <button onClick={() => increasyBy(1)}>+1</button>
                        </>
                        
                    )
                }
            </ProductCard>
        );

        let tree = wrapper.toJSON();
        console.log(tree);
        expect(tree).toMatchSnapshot();

        act(() => {
            (tree as any).children[2].props.onClick();
        })


        tree = wrapper.toJSON();
        //console.log((tree as any).children[1]);
        expect((tree as any).children[1].children[0]).toBe("0")
    })
  
})