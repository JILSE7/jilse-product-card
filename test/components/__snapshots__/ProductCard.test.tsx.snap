// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el product card Debe incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;

exports[`Pruebas en el product card debe mostrar el componente correctamente 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;
